### PR TITLE
Unify isViewable, handle n-dimensional empty tensors.

### PR DESCRIPTION
**Pull request by gchanan**
1) Unifies the two isViewable functions in ATen and TH.
2) Handle n-dimensional empty tensors in the implementation
3) Clarify some comments.

This requires an extra copy in the TH case, but that will go away.

### CODE REVIEW

**Comment by SsnL** [comment]
Can we unify the THC version somehow too? https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensor.cpp#L242

**Comment by gchanan** [comment]
why is there a THC version :(.

**Comment by SsnL** [comment]
Yeah sorry I forgot about that one earlier. :( We basically duplicated everything for TH & THC, so...

**Comment by ezyang** [comment]
We should probably unify THTensor/THCTensor doubletime, but there seem to be some yaks in the way...

File: aten/src/TH/THTensor.cpp [hunk 1]
```diff
@@ -54,3 +56,69 @@ void THTensor_free(THTensor *self)
     }
   }
 }
+
+// On a high level,
+// 1. separate tensor->size into chunks of dimensions, where the dimensions are
+//    ``contiguous'' in each chunk, i.e., stride[i] = size[i+1] * stride[i+1]
+// 2. view_size must be able to be separated into same number of chunks as tensor->size was separated into,
```
Comment (ezyang) [inline aten/src/TH/THTensor.cpp]: Is this comment still valid? No mention of `view_size`.

File: aten/src/TH/THTensor.cpp [hunk 2]
```diff
@@ -54,3 +56,69 @@ void THTensor_free(THTensor *self)
...
+//    where each chunk of view_size has matching ``numel'', i.e., number of subspaces,
+//    as the corresponding chunk of tensor->size.
+at::optional<std::vector<int64_t>>
```
Comment (ezyang) [inline aten/src/TH/THTensor.cpp]: What does the nullopt case signify?

Comment (ezyang) [reply aten/src/TH/THTensor.cpp]: I guess, not possible to restride for the new shape.

**Review by ezyang — APPROVED** [review]
Always hard to tell what has changed when you move code around and make changes at the same time :)

File: aten/src/TH/THTensor.cpp (hunk 2 above)
Comment (gchanan) [reply aten/src/TH/THTensor.cpp]: Right, I just changed it from `{}` to `nullopt` so it's more explicit -- `{}` is a valid initializer for both `optional` and `vector`.

File: aten/src/TH/THTensor.cpp (hunk 1 above)
Comment (gchanan) [reply aten/src/TH/THTensor.cpp]: You are correct, the two implementations used different variable names.

**Pull request merged**